chapters/06_agent_based_extension/01_state_dependent/main.tex

```diff
@@ -0,0 +1,148 @@
+    \caption{Markov chain example with \(C=1, T=1, N=2, M=1\)}
+    \label{fig:state_dependent_markov_example}
+\end{figure}
+
+In the previous chapters of this thesis, the value of the service rate \(\mu\)
+was set to a constant value.
+For this variation of the model \(\mu\) will have a different value depending
+on the number of individuals present in the two nodes of the queue.
+In other words \(\mu\) will take a different value depending on the state
+\(u, v\) of which the system is in.
+Consider the following value of \(\mu\) for the Markov chain model shown in
+Figure~\ref{fig:state_dependent_markov_example}.
+
+\begin{equation*}
+    \mu =
+    \begin{cases}
+        \mu_{(0, 0)}, \text{ if } (u, v) = (0, 0) \\
+        \mu_{(0, 1)}, \text{ if } (u, v) = (0, 1) \\
+        \mu_{(0, 2)}, \text{ if } (u, v) = (0, 2) \\
+        \mu_{(1, 1)}, \text{ if } (u, v) = (1, 1) \\
+        \mu_{(1, 2)}, \text{ if } (u, v) = (1, 2)
+    \end{cases}
+\end{equation*}
+
+
+
+\subsection{Implementation}
+
+
+The state-dependent variation of the queueing system was implemented using the
+python library \texttt{ciw}~\cite{ciwpython}.
+All the code used to implement the state-dependent variation of the queueing
+system is publicly available on GitHub.
+More details on the code can be found in appendix~\ref{appendix:ambulance_game}.
+The \texttt{ciw} library is structured in a way that allows the user to create
+their own class for the service time distribution.
+This class must inherit from the \texttt{ciw.dists.Distribution} class and
+implement the \texttt{\_\_init\_\_} and \texttt{sample} methods.
+The following code shows the implementation of the state-dependent service time
+distribution class.
+
+\begin{lstlisting}[style=pystyle]
+>>> import ciw
+>>> class StateDependentExponential(
+...     ciw.dists.Distribution
+... ):
+...     def __init__(self, rates):
+...         self.rates = rates
+... 
+...     def sample(self, t=None, ind=None):
+...         """
+...         This method is used to sample the service time for an
+...         individual based on the current state
+...         """
+...         state = (
+...             len(ind.simulation.nodes[1].individuals[0]),
+...             len(ind.simulation.nodes[2].individuals[0]),
+...         )
+...         is_invalid_state = (
+...             state[0] > 0 and state[1] < ind.simulation.threshold
+...         )
+...         if is_invalid_state:
+...             state = (state[0] - 1, state[1] + 1)
+...         rate = self.rates[state]
+...         return random.expovariate(rate)
+    
+\end{lstlisting}
+
+The function \texttt{simulate\_model} described in
+Section~\ref{sec:discrete_event_simulation} takes \texttt{mu} as one of its
+arguments.
+Whenever the value of \(\mu\) is a dictionary with keys being the possible
+states and values the service rate for that state, the \texttt{mu} argument
+will be a \texttt{StateDependentExponential} object.
+The following code shows the implementation of the
```

Comment:
Avoid `the following` use `ref` etc...

Reply:
Will do. I don't think I `referece` code snippets throughout the thesis. Will make a new PR to change that everywhere

Reply:
Sorry for missing that before.